quickwit/quickwit-control-plane/src/scheduler.rs

```diff
@@ -17,37 +17,598 @@
+    type ObservableState = IndexingSchedulerState;
 
-    fn observable_state(&self) -> Self::ObservableState {}
+    fn observable_state(&self) -> Self::ObservableState {
+        self.state.clone()
+    }
 
     fn name(&self) -> String {
-        "IndexerScheduler".to_string()
+        "IndexingScheduler".to_string()
     }
 
-    async fn initialize(&mut self, _: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+    async fn initialize(&mut self, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        self.handle(RefreshPlanLoop, ctx).await?;
+        ctx.schedule_self_msg(HEARTBEAT, ControlPlanLoop).await;
         Ok(())
     }
 }
 
 impl IndexingScheduler {
-    pub fn new(cluster: Arc<Cluster>, metastore: Arc<dyn Metastore>) -> Self {
-        Self { cluster, metastore }
+    pub fn new(
+        cluster: Arc<Cluster>,
+        metastore: Arc<dyn Metastore>,
+        indexing_client_pool: IndexingClientPool,
+    ) -> Self {
+        Self {
+            cluster,
+            metastore,
+            indexing_client_pool,
+            state: IndexingSchedulerState::default(),
+        }
+    }
+
+    async fn schedule_indexing_plan(&mut self) -> anyhow::Result<()> {
+        let indexers: Vec<ClusterMember> = self.get_indexers_from_cluster_state().await;
+        if indexers.is_empty() {
+            warn!("No indexer available, cannot build a physical indexing plan.");
+            return Ok(());
+        };
+        let source_configs: HashMap<IndexSourceId, SourceConfig> =
+            self.fetch_source_configs().await?;
+        let indexing_tasks = build_indexing_plan(&indexers, &source_configs);
+        let new_physical_plan =
+            build_physical_indexing_plan(&indexers, &source_configs, indexing_tasks);
+        self.apply_physical_indexing_plan(&indexers, new_physical_plan)
+            .await;
+        Ok(())
+    }
+
+    async fn fetch_source_configs(&self) -> anyhow::Result<HashMap<IndexSourceId, SourceConfig>> {
+        let indexes_metadatas = self.metastore.list_indexes_metadatas().await?;
+        let source_configs: HashMap<IndexSourceId, SourceConfig> = indexes_metadatas
+            .into_iter()
+            .flat_map(|index_metadata| {
+                index_metadata
+                    .sources
+                    .into_iter()
+                    .map(move |(source_id, source_config)| {
+                        (
+                            IndexSourceId {
+                                index_id: index_metadata.index_config.index_id.to_string(),
+                                source_id,
+                            },
+                            source_config,
+                        )
+                    })
+            })
+            .collect();
+        Ok(source_configs)
+    }
+
+    /// Checks if the last applied plan corresponds to the running indexing tasks present in the
+    /// chitchat cluster state. If true, do nothing.
+    /// If false, schedule a new indexing plan.
```

Comment:
Triggering a new schedule plan seems a little bit harsh here.
Can you open a ticket to revisit this logic?

I suspcet it could easily end up in a crazy loop if a single node is faulty, rendering the entire cluster unusablle.
For instance, imagine a node that is missing hard disk.. or a node that cannot connect to kafka. etc.

In fact it is probably worth thinking whetehr an errorr is ndoe speicfic (the node cannot run a task), or source specific (won't work on another node).... Try to infer it... and act differently depending on that.


We could:
a) Send an RPC to try to enforce the plan (also I agree this is strange considering we pass the plan through gRPC already). Afterall. maybe chitchat is just lagging. The node should then tell us if 
it fails to start the pipeline in case we should do b). If we get a success, then we wait.
b)  If the node cannot be accessed, manually marking the faulty node as faulty, putting out of the membership and trigger a replan. This requires refining our membership definition


More generally, my hunch is that we could have decoupled the two following tasks:
- computing a new plan if necessary.
- making sure the plan is applied.